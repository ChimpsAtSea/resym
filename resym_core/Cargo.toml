[package]
name = "resym_core"
version.workspace = true
edition = "2021"
rust-version = "1.75"
exclude = [
    "tests/data/*",
    "tests/snapshots/*",
]

[features]
default = ["rayon", "http"]

rayon = ["dep:rayon", "dep:num_cpus", "dashmap/rayon"]
http = ["dep:ehttp", "dep:url"]

[dependencies]
pdb = { git = "https://github.com/ergrelet/pdb", branch = "0.8.0-resym" }
thiserror = "1.0"
log = "0.4"
rayon = { version = "1.5", optional = true }
num_cpus = { version = "1.13", optional = true }
crossbeam-channel = "0.5"
dashmap = { version = "5.2" }
regex = "1.5"
similar = "2.1"
ehttp = { version = "0.2", optional = true }
url = { version = "2.3", optional = true }

# Web:
[target.'cfg(target_arch = "wasm32")'.dependencies]
wasm_thread = { version = "0.2", features = ["es_modules"] }
instant = { version = "0.1", features = ["wasm-bindgen"] }

[dev-dependencies]
insta = "1.14"
