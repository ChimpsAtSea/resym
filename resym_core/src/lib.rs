pub mod backend;
pub mod diffing;
mod error;
pub mod frontend;
pub mod pdb_file;
pub mod pdb_types;
pub mod syntax_highlighting;

pub use error::*;

const PKG_VERSION: &str = env!("CARGO_PKG_VERSION");

/// Macro used to switch between iterators depending on rayon's availability
#[macro_export]
#[cfg(not(feature = "rayon"))]
macro_rules! cond_par_iter {
    ($expression:expr) => {
        $expression.iter()
    };
}
#[macro_export]
#[cfg(feature = "rayon")]
macro_rules! cond_par_iter {
    ($expression:expr) => {
        $expression.par_iter()
    };
}

/// Macro used to switch between functions depending on rayon's availability
#[macro_export]
#[cfg(not(feature = "rayon"))]
macro_rules! cond_sort_by {
    ($expression:expr, $($x:tt)*) => {
        $expression.sort_by($($x)*)
    };
}
#[macro_export]
#[cfg(feature = "rayon")]
macro_rules! cond_sort_by {
    ($expression:expr, $($x:tt)*) => {
        $expression.par_sort_by($($x)*)
    };
}
