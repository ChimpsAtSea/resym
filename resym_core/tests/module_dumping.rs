use std::path::Path;

use resym_core::{pdb_file::PdbFile, pdb_types::PrimitiveReconstructionFlavor};

const TEST_PDB_FILE_PATH: &str = "tests/data/test.pdb";
const TEST_MODULE_INDEX: usize = 27;
const TEST_MODULE_PATH: &str = "D:\\a\\_work\\1\\s\\Intermediate\\crt\\vcstartup\\build\\xmd\\msvcrt_kernel32\\msvcrt_kernel32.nativeproj\\objd\\amd64\\default_local_stdio_options.obj";

#[test]
fn test_module_dumping_by_path_portable() {
    let pdb_file = PdbFile::load_from_file(Path::new(TEST_PDB_FILE_PATH)).expect("load test.pdb");

    let module_dump = pdb_file
        .reconstruct_module_by_path(TEST_MODULE_PATH, PrimitiveReconstructionFlavor::Portable)
        .unwrap_or_else(|err| panic!("module dumping failed: {err}"));

    insta::assert_snapshot!("module_dumping_by_path_portable", module_dump);
}

#[test]
fn test_module_dumping_by_index_portable() {
    test_module_dumping_by_index_internal(
        "module_dumping_by_index_portable",
        TEST_MODULE_INDEX,
        PrimitiveReconstructionFlavor::Portable,
    );
}

#[test]
fn test_module_dumping_by_index_microsoft() {
    test_module_dumping_by_index_internal(
        "module_dumping_by_index_microsoft",
        TEST_MODULE_INDEX,
        PrimitiveReconstructionFlavor::Microsoft,
    );
}

#[test]
fn test_module_dumping_by_index_raw() {
    test_module_dumping_by_index_internal(
        "module_dumping_by_index_raw",
        TEST_MODULE_INDEX,
        PrimitiveReconstructionFlavor::Raw,
    );
}

fn test_module_dumping_by_index_internal(
    snapshot_name: &str,
    module_index: usize,
    primitives_flavor: PrimitiveReconstructionFlavor,
) {
    let pdb_file = PdbFile::load_from_file(Path::new(TEST_PDB_FILE_PATH)).expect("load test.pdb");

    let module_dump = pdb_file
        .reconstruct_module_by_index(module_index, primitives_flavor)
        .unwrap_or_else(|_| panic!("module dumping"));

    insta::assert_snapshot!(snapshot_name, module_dump);
}
