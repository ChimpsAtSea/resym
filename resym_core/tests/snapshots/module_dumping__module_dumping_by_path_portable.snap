---
source: resym_core/tests/module_dumping.rs
expression: module_dump
---
using namespace std;
using PUWSTR_C = const wchar_t*;
using TP_CALLBACK_ENVIRON_V3 = _TP_CALLBACK_ENVIRON_V3;
uint64_t* (__local_stdio_scanf_options)(); // CodeSize=8
uint64_t _OptionsStorage;
void (__scrt_initialize_default_local_stdio_options)(); // CodeSize=69
